src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            setFirstValue(value);
+            return;
+        }
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+        if (isEmpty()) {
+            setFirstValue(value);
+            return;
+        }
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+        linkBefore(value, findNodeByIndex(index));
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        if (isEmpty()) {
+            setFirstValue(value);
+            return value;
+        }
+        T removedValue = remove(index);
+        add(value, index);
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        T value = node.item;
+        unlink(node);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        try {
+            unlink(findNodeByValue(object));
+        } catch (RuntimeException e) {
+            return false;
+        }
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void checkAddIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index was " + index
+                    + ", but size is " + size);
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index was " + index
+                    + ", but size is " + size);
+        }
+    }
+
+    private Node<T> findNodeByValue(T value) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if (value == node.item || value != null
+                    && value.equals(node.item)) {
+                return node;
+            }
+            node = node.next;
+        }
+        throw new RuntimeException("Element didn't find");
+    }
```

Comment:
if you can find node by index, you don't need to finding by value

Reply:
I need the logic of finding the value anyway, so I decided to create separate method for this